packages/react-union/src/components/Union/Union.js

```diff
@@ -45,44 +49,58 @@ class Union extends Component {
 	};
 
 	static defaultProps = {
+		isServer: false,
 		onScanEnd: noop,
 		onScanError: noop,
 		onScanStart: noop,
+		parent: document && document.body,
 		strictMode: true,
 	};
 
-	state = {
-		configs: [],
-	};
+	static scan = props => {
+		const { onScanStart, onScanEnd, onScanError, parent, routes } = props;
 
-	componentDidMount() {
-		this.scan(this.props);
-	}
+		try {
+			onScanStart();
+			const scanningResult = scan(routes, parent);
+			onScanEnd(scanningResult);
 
-	componentDidUpdate(prevProps) {
-		if (prevProps.routes !== this.props.routes) {
-			this.scan(prevProps);
-		}
-	}
+			return scanningResult;
+		} catch (error) {
+			onScanError(error);
 
-	scan = props => {
-		const { onScanStart, onScanEnd, onScanError, parent, routes } = props;
+			throw error;
+		}
+	};
 
-		onScanStart();
+	static getDerivedStateFromProps(nextProps, prevState) {
+		if (prevState.routes !== nextProps.routes) {
+			return {
+				...Union.scan(nextProps),
+				routes: nextProps.routes,
+			};
+		}
 
-		const domParent = parent || document.body;
+		return null;
+	}
 
-		scan(routes, domParent).then(
-			configs => {
-				onScanEnd(configs);
-				this.setState({ configs });
-			},
-			error => onScanError(error)
-		);
+	state = {
+		...(this.props.isServer ? Union.scan(this.props) : { commonData: {}, configs: [] }),
```

Comment:
This looks ugly. I will wrap this in a `scanResult` property instead. It will also make the state structure more clear.